consensus/common/src/main/java/tech/pegasys/pantheon/consensus/common/jsonrpc/AbstractGetSignerMetricsMethod.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.common.jsonrpc;
+
+import tech.pegasys.pantheon.consensus.common.VoteTallyCache;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.JsonRpcRequest;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.parameters.JsonRpcParameter;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.queries.BlockchainQueries;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcError;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcErrorResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.ProposerReportBlockProductionResult;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.SignerMetricsResult;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.results.report.block.production.ValidatorReportBlockProductionResult;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.LongStream;
+
+public abstract class AbstractGetSignerMetricsMethod {
+
+  private final BlockchainQueries blockchainQueries;
+  private final VoteTallyCache voteTallyCache;
+  private final JsonRpcParameter parameters;
+
+  public AbstractGetSignerMetricsMethod(
+      final BlockchainQueries blockchainQueries,
+      final VoteTallyCache voteTallyCache,
+      final JsonRpcParameter parameter) {
+    this.blockchainQueries = blockchainQueries;
+    this.voteTallyCache = voteTallyCache;
+    this.parameters = parameter;
+  }
+
+  public JsonRpcResponse response(final JsonRpcRequest request) {
+
+    final long startBlock = parameters.required(request.getParams(), 0, Long.class);
+    final long endBlock = parameters.required(request.getParams(), 1, Long.class);
+
+    if (isValidParameters(startBlock, endBlock)) {
+
+      final Map<Address, ProposerReportBlockProductionResult> proposersMap = new HashMap<>();
+      final Map<Address, ValidatorReportBlockProductionResult> validatorInLastBlockMap =
+          new HashMap<>();
+
+      // go through each block (startBlock is inclusive and endBlock is exclusive)
+      LongStream.range(startBlock, endBlock)
+          .boxed()
+          .sorted(Collections.reverseOrder())
+          .forEach(
```

Comment:
It's actually simpler and probably faster to iterate this from start to end block so you can just use `LongStream.range(startBlock, endBlock).forEach(...)`.

Going forwards through the range makes more sense because the list of validators is adjusted based on votes from  in the block so you have to process blocks in order to know who the current list of validators is.  Pantheon has some caching for those results so we don't always have to reprocess a lot of blocks, but if it's a big range it will be a lot faster to go forwards than backwards because it works better with that caching.